File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
vector<vector<int>>ans;
    void solve(int indx,int n,vector<int>&arr,vector<int>&empty,int k){
        if(k==0){
            ans.push_back(empty);
            return;
            }
        if(indx>=n || k<0){
            return;
        }
        empty.push_back(arr[indx]);
        solve(indx,n,arr,empty,k-arr[indx]);
        empty.pop_back();
        solve(indx+1,n,arr,empty,k);
    }
    vector<vector<int>> combinationSum(vector<int>& arr, int target) {
        vector<int>empty;
        int n=arr.size();
        solve(0,n,arr,empty,target);
        return ans;


        
    }
};

File: 39-combination-sum/combination-sum.java
class Solution {
    List<List<Integer>> result = new ArrayList<>();
    public void solve(int i,int[] arr,ArrayList<Integer> empty,int target){
        if(target==0){
            result.add(new ArrayList<>(empty));
            return;
        }
        if(i>=arr.length || target<0){
            return;
        }
        empty.add(arr[i]);
        solve(i,arr,empty,target-arr[i]);
        empty.remove(empty.size()-1);
        solve(i+1,arr,empty,target);

    }
    public List<List<Integer>> combinationSum(int[] arr, int target) {
        int n=arr.length;
        ArrayList<Integer> empty = new ArrayList<>();
        solve(0,arr,empty,target);
        return result;
    }
}